Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php

```diff
@@ -230,18 +231,22 @@ private function whenRootWorkspaceWasCreated(RootWorkspaceWasCreated $event): vo
         }
     }
 
-    private function whenNodeAggregateWasMoved(NodeAggregateWasMoved $event): void
+    private function whenNodeAggregateWasMoved(NodeAggregateWasMoved $event, EventEnvelope $eventEnvelope): void
     {
-        // WORKAROUND: we simply use the first MoveNodeMapping here to find the dimension space point
-        // @todo properly handle this
-        /* @var \Neos\ContentRepository\Core\Feature\NodeMove\Dto\OriginNodeMoveMapping[] $mapping */
-        $mapping = iterator_to_array($event->nodeMoveMappings);
-
-        $this->markAsMoved(
-            $event->getContentStreamId(),
-            $event->getNodeAggregateId(),
-            $mapping[0]->movedNodeOrigin
-        );
+        // Changes reflect the editorial intention, not the effect as they are later published via commands
+        $metadata = $eventEnvelope->event->metadata?->value ?? [];
+        if (isset($metadata['commandPayload'])) {
+            $command = MoveNodeAggregate::fromArray($metadata['commandPayload']);
+            // WORKAROUND: we simply use the command's DSP here as the origin dimension space point.
```

Comment:
As far as I know we only relied on the command-metadata in the write side for rebasing, right?
I still hope that we can replace that at some point with a pure event based approach.
I don't fully understand the change yet, but I would really try to keep that implementation detail as local as possible and it makes me think: Do we maybe miss properties in the event if we have to re-create the command at projection time?

Reply:
I think the issue here is with the change projection. If it needs to know about origin DSPs it has to keep track of them itself, then the event will suffice again. Then again, I don't really know what the change projection does with MoveNode anyway. This is to be discussed elsewhere and I am pretty sure this workaround can then be removed.

Reply:
this is an interesting question, for whom do we make the events, something we should _really_ define for the future as well. As in do we record only whatever at that point is relevant to the content graph or anything that might be of interested at this particular point?

Reply:
I need to understand this case better – but in general I would say that we consider the "metadata" of events internal

Reply:
We don't make the events for anyone, we just need to communicate the effect of what happened. And I'm pretty sure that's what we do ^^
If reading the metadata is the issue we can as well use the first affected DSP instead of the explicitly selected one, should be the same.